### PR TITLE
add library AP_SBusOut

### PR DESCRIPTION
**Pull request by kd0aij**
supersedes #6794

Adds support for SBUS serial output in AP_HAL for all vehicle types.

- This PR implements the UARTDriver configure_parity and set_stop_bits methods in AP_HAL_PX4 and AP_HAL_SITL.

- Tested on SITL using Prolific Technology, Inc. PL2303 Serial Port on USB.

- Flight tested on px4-v4 (PixRacer) with ArduPlane using FrSky SBUS-PWM converter on Serial2 (telem2, SERIAL2_PROTOCOL=15).

- selected UART must support baud rate of 100,000, even parity and 2 stop bits

- to implement for other board types, provide iplementations for the default empty methods in AP_HAL/UARTDriver.h; providing an unbuffered write method will reduce timing jitter on the sbus output frames

### CODE REVIEW

**Comment by kd0aij** [comment]
@tridge Can't figure out why I get this error:
```
lib/libArduPlane_libs.a(SRV_Channels.cpp.0.o): In function `SRV_Channels::SRV_Channels()':
SRV_Channels.cpp:(.text._ZN12SRV_ChannelsC2Ev+0x3f): undefined reference to `AP_SBusOut::AP_SBusOut()'
```

File: libraries/AP_SBusOut/AP_SBusOut.cpp [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+#define SBUS_SCALE (2048.0f / (SBUS_MAX - SBUS_MIN))
+
+const AP_Param::GroupInfo AP_SBusOut::var_info[] = {
+    // @Param: SBUS_RATE
+    // @DisplayName: SBUS default output rate
+    // @Description: This sets the SBUS output frame rate in Hz.
+    // @Range: 25 250
+    // @User: Advanced
+    // @Units: Hz
+    AP_GROUPINFO("RATE",  19, AP_SBusOut, sbus_rate, 50),
```
Comment (tridge) [inline libraries/AP_SBusOut/AP_SBusOut.cpp]: starting with index 1 would be better

File: libraries/SRV_Channel/SRV_Channels.cpp [hunk 2]
```diff
@@ -117,6 +118,10 @@ const AP_Param::GroupInfo SRV_Channels::var_info[] = {
     // @Path: ../AP_Volz_Protocol/AP_Volz_Protocol.cpp
     AP_SUBGROUPINFO(volz, "_VOLZ_",  19, SRV_Channels, AP_Volz_Protocol),
 
+    // @Group: SBUS_
+    // @Path: ../AP_SBusOut/AP_SBusOut.cpp
+    AP_SUBGROUPINFO(volz, "_SBUS_",  20, SRV_Channels, AP_SBusOut),
```
Comment (tridge) [inline libraries/SRV_Channel/SRV_Channels.cpp]: should be sbus, not volz

**Comment by kd0aij** [comment]
@tridge SITL log showing elevon and tilt PWM outputs changing during bootup (I think)
This is apparent on the Pixracer when booting with the safety switch disabled.

**Comment by kd0aij** [comment]
log of servo outputs from Pixracer with BRD_SAFETYENABLE=0
RCOU.Cn are scaled by .01 (scaled range [100,200])
elevons are channels 2,4, starting PWM values are near 1900
vehicle is stationary until about 38 seconds
elevons take about 35 seconds to settle down

log_px4_startup

**Comment by kd0aij** [comment]
Qmodes flight test log: https://drive.google.com/open?id=1HtSo643QU8Sb3HHBhSYAf41F2H-_ca0h
rollpitchtracking

**Comment by kd0aij** [comment]
@samuelctabor This is an improved version of #6794 
It would be great if you could test it on PX4

**Comment by kd0aij** [comment]
flight test with manual tilt bias addition: https://github.com/ArduPilot/ardupilot/pull/7294#issue-276895803

**Comment by samuelctabor** [comment]
Hi @kd0aij 
I'm testing this just now. I also have a FrSky SBus to PWM converter. How should I hook it up? Power from the servo rail, and connect TX&RX to SBus signal line? Any inverter required?

Cheers, Sam

**Comment by kd0aij** [comment]
Yes, I used an inverter on the Telem2 TX pin; schematic is in the blog here: https://discuss.ardupilot.org/t/e-flite-convergence-with-pixracer-and-sbus-output/21999
SBUS1 is output only; Telem2 RX pin is not connected.

I used an ACSP7 power module for the PixRacer and a separate BEC to power the servo rail.

**Comment by samuelctabor** [comment]
Thanks. I've connected with an inverter but I'm not getting clean packets through - the light on the converter is sporadically flashing and the servo is twitching. I'm testing on ArduPlane and I've set the parameters 

SERIAL2_BAUD     100.000000
SERIAL2_PROTOCOL 15.000000

In the GCS I get a few SBus related messages:

Init Gyro***
online system 1
Mode INITIALISING
fence present
Vcc 4.3
AP_SBusOut: init 50 Hz
channel 9 pwm: 1500
APM: Throttle failsafe on
APM: Failsafe. Short event on: type=1/reason=3
APM: Flight mode = 1

And just to confirm, it's attached to the 6-pin JST-GH beside the USB connector.

**Comment by kd0aij** [comment]
I assume you've programmed the SBUS adapter for channels 9,10 and 11?

**Comment by samuelctabor** [comment]
No I have it configured for channels 1-4, I assumed all 16 channels would be sent out over SBus?

**Comment by kd0aij** [comment]
yes, all 16 channels go out, but I used 1-4 for servos at 70Hz and 9-11 for ESCs on SBUS since my adapter was generating a PWM rate of 168 Hz regardless of SBUS frame rate. Are you also seeing 168Hz on the SBUS adapter PWM out?

**Comment by samuelctabor** [comment]
I can't say for sure. I have tested the adaptor taking input direct from a Frsky X4R receiver with micro servos on the outputs, this works fine. Then I am essentially just adding the Pixracer and inverter in between.

**Comment by kd0aij** [comment]
image

I think you're on Telem1, not Telem2

**Comment by samuelctabor** [comment]
Well spotted, I was going by the pin names not the silkscreen. It's now working on the bench, should be able to do a flight test at the weekend. Thanks for your help!

**Comment by kd0aij** [comment]
Great, glad it's working for you.

**Comment by kd0aij** [comment]
@tridge I minimized the change to the PX4 UARTDriver:write method as requested, turned off the debug flag in AP_SBusOut and bench tested; looks good to me now.

**Comment by kd0aij** [comment]
jitter measurement with 200Hz frame rate on px4-v4 (from #6794):
Using AUX6 as a gpio toggled across the call to _write_fd(), observed max latency to uart start bit is 231 usec and max time in _write_fd() is 304 usec. (frame rate 200Hz)

Implication is that sbus output frame timing jitter is < 300 usec with this implementation, and that an interframe gap on the order of 2 msec should be reliable. 

Min and max sbus frame intervals were 4.58 and 5.42 msec and the minimum frame gap was 0.991 msec on the 10 second test interval.

sbus_latency_200hz

histogram of sbus frame intervals shows that jitter is reasonable:

sbus_interval_histo

**Pull request merged**